src/bikeracks/votes.py

```diff
@@ -31,133 +31,140 @@ def get_vote_data(rack_id, user_id):
-                    h.decrement_downvote_count(rack_id, db)
-                    
-                # if the user downvoted
-                elif vote_type == -1:
-                    # update the upvote_count in bikeracks for this rack, decrement by 1
-                    h.decrement_upvote_count(rack_id, db)
-                
-            else:
-                # voting for the first time
-                
-                query = "INSERT INTO votes (rack_id, user_id, vote_type) VALUES (?, ?, ?)"
-                db.execute(query, (rack_id, user_id, vote_type))
-                db.commit()
-                
-            # increment the upvote or downvote count
+            query = "UPDATE votes SET vote_type = ? WHERE rack_id = ? AND user_id = ?"
+            db.execute(query, (vote_type, rack_id, user_id))
+            db.commit()
+            # also update the vote count in bikeracks table
             if vote_type == 1:
-                # update the upvote_count in bikeracks for this rack, increment by 1
-                h.increment_upvote_count(rack_id, db)
+                # update the downvote_count in bikeracks for this rack, decrement by 1
+                h.decrement_downvote_count(rack_id, db)
+                
+            # if the user downvoted
             elif vote_type == -1:
-                # update the downvote_count in bikeracks for this rack, increment by 1
-                h.increment_downvote_count(rack_id, db)
+                # update the upvote_count in bikeracks for this rack, decrement by 1
+                h.decrement_upvote_count(rack_id, db)
+            
+        else:
+            # voting for the first time
             
-            resp = get_vote_data(rack_id, user_id)
-            return jsonify(resp)
-        except sqlite3.Error as e:
-            print("Database error:", e)
-            return "", 500
-        except KeyError as key_e:
-            print("Error with key: {}".format(key_e))
+            query = "INSERT INTO votes (rack_id, user_id, vote_type) VALUES (?, ?, ?)"
+            db.execute(query, (rack_id, user_id, vote_type))
+            db.commit()
+            
+        # increment the upvote or downvote count
+        if vote_type == 1:
+            # update the upvote_count in bikeracks for this rack, increment by 1
+            h.increment_upvote_count(rack_id, db)
+        elif vote_type == -1:
+            # update the downvote_count in bikeracks for this rack, increment by 1
+            h.increment_downvote_count(rack_id, db)
+        
+        resp = get_vote_data(rack_id, user_id)
+        return jsonify(resp)
+    except sqlite3.Error as e:
+        print("Database error:", e)
+        return "", 500
+    except KeyError as key_e:
+        print("Error with key: {}".format(key_e))
+        return "", 500
     # TODO do I actually need below line
-    # maybe return 500 here
+   # maybe return 500 here
     return "", 500
     
 @votes.route('/unvote', methods=['POST'])
```

Comment:
Seems like this shouldn't need to be its own endpoint separate from `/submit_vote`.

You could using an HTTP `DELETE` request. If you do, then you'll need to make multiple queries to the db so put them in a [single transaction](https://stackoverflow.com/a/19524679) in case something goes wrong.

```python
@votes.route('/submit_vote, methods=['POST', 'DELETE'])
def submit_vote(..):
  upvote_delta = ...
  downvote_delta = ...
  with get_db() as cursor:
    if request.method == 'DELETE':
      # remove row from votes
      cursor.execute('DELETE .... ')  
    # update counts in bikeracks
      cursor.execute('INSERT INTO bikeracks....'.format(up=upvote_delta, down=downvote_delta))

Reply:
I feel like this needs it's own pull request. also i don't really understand.

Reply:
Maybe put it in an issue.

Reply:
Agreed this should be a separate issue. Filed #83.